server/task/src/main/java/org/apache/james/task/MemoryTaskManagerWorker.java

```diff
@@ -0,0 +1,139 @@
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+import org.apache.james.util.FunctionalUtils;
+import org.apache.james.util.MDCBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+public class MemoryTaskManagerWorker implements TaskManagerWorker {
+    private static final boolean INTERRUPT_IF_RUNNING = true;
+    private static final Logger LOGGER = LoggerFactory.getLogger(MemoryTaskManagerWorker.class);
+    public static final Duration CHECK_CANCELED_PERIOD = Duration.ofMillis(100);
+    public static final int FIRST = 1;
+    private final ConcurrentHashMap<TaskId, CompletableFuture<Task.Result>> idToFuture = new ConcurrentHashMap<>();
+
+    @Override
+    public Mono<Task.Result> executeTask(TaskWithId taskWithId, Consumer<TaskExecutionDetailsUpdater> updateDetails) {
+        CompletableFuture<Task.Result> futureResult = CompletableFuture.supplyAsync(() -> runWithMdc(taskWithId, updateDetails));
+
+        idToFuture.put(taskWithId.getId(), futureResult);
+
+        Mono<Task.Result> result = Mono.<Task.Result>fromFuture(futureResult)
+            .doOnError(res -> {
+                if (!(res instanceof CancellationException)) {
+                    failed(updateDetails,
+                        (logger, details) -> logger.error("Task was partially performed. Check logs for more details"));
+                }
+            })
+            .doOnTerminate(() -> idToFuture.remove(taskWithId.getId()));
+
+        return result;
+    }
+
+    private Task.Result runWithMdc(TaskWithId taskWithId, Consumer<TaskExecutionDetailsUpdater> updateDetails) {
+        return MDCBuilder.withMdc(
+            MDCBuilder.create()
+                .addContext(Task.TASK_ID, taskWithId.getId())
+                .addContext(Task.TASK_TYPE, taskWithId.getTask().type())
+                .addContext(Task.TASK_DETAILS, taskWithId.getTask().details()),
+            () -> run(taskWithId, updateDetails));
+    }
+
+
+    private Task.Result run(TaskWithId taskWithId, Consumer<TaskExecutionDetailsUpdater> updateDetails) {
+        updateDetails.accept(TaskExecutionDetails::start);
+        try {
+            return taskWithId.getTask()
+                .run()
+                .onComplete(() -> success(updateDetails))
+                .onFailure(() -> failed(updateDetails, (logger, details) -> logger.error("Task was partially performed. Check logs for more details" + details.getTaskId())));
+        } catch (Exception e) {
+            failed(updateDetails,
+                (logger, executionDetails) -> logger.error("Error while running task", executionDetails, e));
+            return Task.Result.PARTIAL;
+        }
+    }
+
+    @Override
+    public void cancelTask(TaskId id, Consumer<TaskExecutionDetailsUpdater> updateDetails) {
```

Comment:
This method is clearly too long (26 lines) and too complex (6 level of indentation)

Can we consider extracting things in submethods?